lib/posix/options/env.c

```diff
@@ -0,0 +1,263 @@
+	}
+
+	return ret;
+}
+
+int setenv(const char *name, const char *val, int overwrite)
+{
+	int ret = 0;
+	char *env;
+	char **envp;
+	size_t esize;
+	const size_t vsize = (val == NULL) ? 0 : strlen(val);
+	const size_t nsize = (name == NULL) ? 0 : strlen(name);
+	/* total size of name + '=' + val + '\0' */
+	const size_t tsize = nsize + 1 /* '=' */ + vsize + 1 /* '\0' */;
+
+	if (name == NULL || val == NULL) {
+		LOG_DBG("Invalid name '%s' or value '%s'", name, val);
+		errno = EINVAL;
+		return -1;
+	}
+
+	K_SPINLOCK(&environ_lock)
+	{
+		ret = findenv(name, nsize);
+		if (ret == -EINVAL) {
+			LOG_DBG("Invalid name '%s'", name);
+			K_SPINLOCK_BREAK;
+		}
+		if (ret >= 0) {
+			/* name was found in environ */
+			esize = strlen(environ[ret]) + 1;
+			if (overwrite == 0) {
+				LOG_DBG("Found entry %s", environ[ret]);
+				ret = 0;
+				K_SPINLOCK_BREAK;
+			}
+		} else {
+			/* name was not found in environ -> add new entry */
+			esize = environ_size();
+			envp = realloc(environ, sizeof(char **) *
+							(esize + 1 /* new entry */ + 1 /* NULL */));
+			if (envp == NULL) {
+				ret = -ENOMEM;
+				K_SPINLOCK_BREAK;
+			}
+
+			if (TRACK_ALLOC) {
+				allocated += sizeof(char **) * (esize + 2);
+				LOG_DBG("realloc %zu bytes (allocated: %zu)",
+					sizeof(char **) * (esize + 2), allocated);
+			}
+
+			environ = envp;
+			ret = esize;
+			environ[ret] = NULL;
+			environ[ret + 1] = NULL;
+			esize = 0;
+		}
+
+		if (esize < tsize) {
+			/* need to malloc or realloc space for new environ entry */
+			env = realloc(environ[ret], tsize);
+			if (env == NULL) {
+				ret = -ENOMEM;
+				K_SPINLOCK_BREAK;
+			}
+			if (TRACK_ALLOC) {
+				allocated += tsize - esize;
+				LOG_DBG("realloc %zu bytes (allocated: %zu)", tsize - esize,
+					allocated);
+			}
+			environ[ret] = env;
+		}
+
+		strncpy(environ[ret], name, nsize);
+		environ[ret][nsize] = '=';
+		strncpy(environ[ret] + nsize + 1, val, vsize + 1);
+		LOG_DBG("Added entry %s", environ[ret]);
+
+		ret = 0;
+	}
+
+	if (ret < 0) {
+		errno = -ret;
+		ret = -1;
+	}
+
+	return ret;
+}
+
+int unsetenv(const char *name)
+{
+	int ret = 0;
+	size_t esize;
+	size_t nsize;
+
+	nsize = (name == NULL) ? 0 : strlen(name);
+	K_SPINLOCK(&environ_lock)
+	{
+		ret = findenv(name, nsize);
+		if (ret < 0) {
+			ret = (ret == -EINVAL) ? -EINVAL : 0;
+			K_SPINLOCK_BREAK;
+		}
+
+		esize = environ_size();
+		if (TRACK_ALLOC) {
+			allocated -= strlen(environ[ret]) + 1;
+			LOG_DBG("free %zu bytes (allocated: %zu)", strlen(environ[ret]) + 1,
+				allocated);
+		}
+		free(environ[ret]);
+
+		/* shuffle remaining environment variable pointers forward */
+		for (; ret < esize; ++ret) {
+			environ[ret] = environ[ret + 1];
+		}
+		/* environ must be terminated with a NULL pointer */
+		environ[ret] = NULL;
+
+		/* reduce environ size and update allocation */
+		--esize;
+		if (esize == 0) {
+			free(environ);
+			environ = NULL;
+		} else {
+			environ = realloc(environ, (esize + 1 /* NULL */) * sizeof(char **));
```

Comment:
Looks like a possible memory leak here. If realloc returns null we loose the original content of environ

Reply:
Fixed!